pypesto/objective.py

```diff
@@ -188,38 +213,195 @@ def call_mode_res(self, x, sensi_orders):
             else:
                 res = self.res(x)
                 sres = self.sres(x)
+            if self.trace is not None:
+                self.update_trace(np.power(res, 2).sum(), x)
             return res, sres
         else:
             raise ValueError("These sensitivity orders are not supported.")
 
+    def reset_history(self,
+                      temp_file=None,
+                      temp_save_iter=10):
+        """
+        Method to reset the evaluation history of the objective and specify
+        temporary saving options.
+
+        Parameters
+        ----------
+
+        temp_file: filename
+            If specified, temporary results of traces for every optimization
+            run will be saved to that file
+
+        temp_save_iter: int
+            Update interval for temporary saving.
+        """
+
+        self.n_fval = 0
+        self.n_grad = 0
+        self.n_hess = 0
+        self.min_fval = float('Inf')
+        self.start_time = time.time()
+        self.min_x = None
+
+        parameter_names = self.get_parameter_names()
+
+        if parameter_names is None:
+            parameter_names = ['x' + str(ix) for ix in range(self.dim)]
+        else:
+            if len(parameter_names) != self.dim:
+                raise ValueError('List of parameter names must be of the same'
+                                 'length as number of parameters specified'
+                                 'at initialization of the objective')
+
+        if temp_file is not None:
+            cols = ['time', 'n_fval', 'n_grad', 'n_hess', 'fval'] \
+                + parameter_names
+            self.trace = pd.DataFrame(columns=cols)
+            self.temp_file = temp_file
+            self.temp_save_iter = temp_save_iter
+        else:
+            self.trace = None
+            self.temp_file = None
+            self.temp_save_iter = None
+
+    def update_eval_counts(self, sensi_orders):
```

Comment:
nice solution. i was wondering whether it would be possible to move this "counting stuff" to a wrapper class, so as to disentangle the objective and gradient computation from the counting part. the wrapping could then occur before every local optimization, also helping avoid parallelization and history issues.
but that can still be done later.

Reply:
Sure, we could push all the newly introduced fields to a different class but I do not see a way around attaching that class to the objective. But maybe there is some python magic that I do not know of.

Regarding parallelization, I think really the best way to do this is have pypesto being started from separate jobs. This is the only way to effectively parallelize across nodes on a cluster and also works on a local machine.
It would not work with cooperative scatter search strategies, but there we probably want to introduce an additional level of hierarchy anyways.

Reply:
Jan also mentioned that we would like to have transparent parallelization engines in optimization and sampling à-la pyabc using e.g. redis-servers, much of which one might be able to re-use. in optimization first of all I guess only starts would be parallelized, but also startpoint selection.

Reply:
> what do you mean by separate jobs?

submitting individual batch jobs for every individual (subbatch of) optimization(s). e.g. using snakemake:

```
rule parameter_estimation:
	input:
		model=rules.build_amici.output,
		data=rules.process_data.output,
		script='parameter_estimation.py',
	output:
		os.path.join('results', 'multistart', '{modelname}', 'result_{dataset}_{repeat}.pickle')
	wildcard_constraints:
		modelname='[\w\d_]+',
		dataset='[\w\d]+',
		repeat='\d+'
	shell:
		expand('python3 {{input.script}} {{input.model}} {{wildcards.dataset}} {{wildcards.repeat}} {n_starts}', n_starts=MS_PER_JOB)[0]
```

This should fit with the use of redis servers.